### ISSUE TITLE
Investigate potential for using shared mutations for clustering

### ISSUE THREAD

**Issue by matsen** [psathyrella/partis#175]
[This paper](http://dx.doi.org/10.1098/rstb.2014.0242) finds that B cell phylogenies look like this:

![screenshot 2016-03-23 at 6 56 11 am](https://cloud.githubusercontent.com/assets/112708/13987416/af4fc1ac-f0c4-11e5-8564-e63b62467c3c.png)

This could be very helpful for BCR clustering, because we could use shared mutations to distinguish between lineages.

In a collection of the biggish clusters on real data, do we see this too? 

This could be evaluated by either building trees or by looking at shared mutations in germline encoded regions.

**Comment by psathyrella** [comment]
Very similar to #29

**Comment by matsen** [comment]
Make this a second phase?

**Comment by psathyrella** [comment]
Resurrecting this as per discussion yesterday.

I never got around to this in the past largely I think because it seemed difficult to be principled about how much shared mutation to require without making assumptions about tree shape that would be hard to justify. But it definitely would've always been _ nice _ to have it, since we've for sure always over merged light chain, as detailed for instance [here](https://github.com/matsengrp/loris-experiments/issues/224), and splitting families without shared mutation would for sure mostly remove that. 

I remember this [spisak paper](https://doi.org/10.1101/2022.12.22.521661) seemed, on cursory reading, to do something sensible with shared mutation, and I was interested in importing the code and integrating with partis, but the code wasn't really usable when I tried. They subsequentlly updated the repo, and I can't find notes about having tried again after that, so it might now be feasible to import/run.

Yesterday I was saying maybe we could just split subclusters with zero shared mutations. This could work, but we have to remember that not all called mutations are real. Nevertheless, I think this quick heuristic could be the way to go. I think we do need something along these lines sooner than later, though, it's becoming a big problem to just overmerge everything in these samples.

@matsen

**Comment by psathyrella** [comment]
Just for a little more intuition, this is the typical overmerged light chain family that we're dealing with:

![Image](https://github.com/user-attachments/assets/5a9ea8f0-8a9d-4ea3-9869-06c8a4e146dc)

The much higher and concentrated SHM in cdr3 compared to outside of it is pretty definitive -- obviously that _ could _ happen, but it would be very unlikely. So an additional source of information that we could use is SHM uniformity. This would be quite similar to the chimera detection that we've discussed elsewhere, but basically when grouped into real families, sequences should have continously-varying SHM density (average SHM in CDR3 is ~double elsewhere, but still shouldn't really target e.g. four or five consecutive bases as here).

**Comment by matsen** [comment]
My main thought here is that this isn't a quick fix but a big project. Let's start by running the Hilary software and see how it goes. Seem reasonable?

**Comment by psathyrella** [comment]
You're definitely right that it could be a big project, but I think it's also possible that we could make it a quick fix. Requiring one or two shared mutations among sequences in a clonal family would for sure be easy to implement, and it's also very easy to run a new method on the paired clustering paper samples. If it works well and we don't get sucked into thinking about a principled approach, that would likely be a few days of work. To me there's a big difference between trying to do optimal clustering using this new information source, vs doing the simplest possible thing to probably largely ameliorate a big problem with our current clustering.
  
But, yeah, potentially risky because if it kind of works we'll be tempted to just look a little harder. I'm happy to push this off for a while, but I don't think we should wait too long.

**Comment by matsen** [comment]
Up to you! I'd run Hilary. Running other software is easy with CC now.

**Comment by psathyrella** [comment]
Yeah, I'd for sure run Hilary first when I get to this. The thing that may be easier about the non-hilary option, though, is that hilary would give us a full/most-likely partition, and combining that with an existing partition might be challenging (assuming, as I seem to recall from when I reviewed their paper, that we can't just use hilary clustering). Whereas the one-shared-mutation or similar approach would be just a minimal list of splits that we're totally sure of that would be easy to apply to an existing (overmerged) partition. (edit: ah [found it](https://github.com/psathyrella/talks/tree/main/23-spisak-simu), see conclusion slide for tldr). Anyway, hopefully hilary is fast and has an adjustable threshold so we can easily integrate it.

As long as I'm cleaning up my todo file, I also had written down to think about using the ideas from the [Quake/grmlin germline inference method](https://www.biorxiv.org/content/10.1101/2023.11.25.568681v1) to help with shared SHM clustering. Unfortunately I wrote these notes a long time ago and despite re skimming the paper I'm not quite sure what I meant, but I think my basic idea was that they've uncovered a real new source of information by making this 2d plot in panel b:

![Image](https://github.com/user-attachments/assets/65187040-30fb-48a5-9d79-98f834bbea2c)
  
Fig 1 is also good for context:

![Image](https://github.com/user-attachments/assets/70cd9aee-12cc-4c00-8f6d-3af97fdbd7ea)
